### ISSUE TITLE
Documenting Fields?

### ISSUE THREAD

**Issue by AnthonySuper** [morpheusgraphql/morpheus-graphql#404]
Is there a plan to implement field-level documentation?

It's not super typesafe, but honestly providing a function of `String -> Maybe Text` that is called with each field inside the `GQLType` class could work.

**Comment by nalchevanidze** [comment]
if you are using template haskell for gql types then you can document types with gql syntax.

schema.gql
```gql
"""
Description for Deity
"""
type Deity {
  """
  Description for name
  """
  name: String!
  power: String @deprecated(reason: "some reason for")
}
```

another way with just strings is unsafe. but  there is one way that could work.

you could try it with template haskell. i don't know if it works. 

idea is to define function that takes template haskell names and converts it to string for introspection.

```hs
fieldDescriptions ::  Proxy a -> [(TH.Name,Text)]
```

```hs
{-# LANGUAGE TemplateHaskell   #-}

data Deity = Deity
  { fullName ::  Text
  , power    :: Maybe Text 
  } deriving (Generic,GQLType)

instance GQLType Deity where 
    fieldDescriptions = [
       ('fullName, "Description for name")
     ]
````

**Comment by AnthonySuper** [comment]
Using Template Haskell names for labels would work, but I don't think that really gains much safety as there's no association between the label and the parsed type, so it's basically like using a string anyway. I think that it would be okay to be not-as-typesafe because we're just doing documentation, and the only way to do this type-safely is probably some super gross higher-kinded-data thing that would make writing the rest of the application very annoying.

**Comment by AnthonySuper** [comment]
Actually, if type safety is desired, it is probably possible to write something like this with GADTs and what `Generic` gives you:

```haskell
data FieldTag (struct :: Type) (tag :: Type) where 
  FieldTag :: (HasField field symbol, KnownSymbol field) => Proxy field -> tag -> FieldTag struct field tag
  -- ^ Use something similar to HasField from https://hackage.haskell.org/package/generic-lens-2.0.0.0/docs/Data-Generics-Product-Fields.html
```

Then the field names would have type `[FieldTag Diety Text]` or similar. But this is kinda hard to use, maybe.

**Comment by nalchevanidze** [comment]
as long description is not security relevant. api can be like:

```hs
fieldDescriptions ::  Proxy a -> [(Text,Text)]
```

**Comment by nalchevanidze** [comment]
implemented in: #513

```hs
getDescriptions :: f a -> Map Text Description
  getDescriptions _ = mempty

**Comment by nalchevanidze** [comment]
@AnthonySuper please close this. issue if it works for you
